pkg/utils/requiredfieldsvalidator/validator.go

```diff
@@ -0,0 +1,139 @@
+package requiredfieldsvalidator
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/instaclustr/operator/pkg/models"
+)
+
+func ValidateRequiredFields(v any) error {
+	val := reflect.ValueOf(v)
+
+	if val.Kind() == reflect.Ptr {
+		val = val.Elem()
+	}
+
+	if val.Kind() != reflect.Struct {
+		return fmt.Errorf("expected struct type - got the %s", val.Kind())
+	}
+
+	err := validateValue(val)
+	if err != nil {
+		return fmt.Errorf("error have occured: %w", err)
+	}
+
+	return nil
+}
+
+func validateValue(val reflect.Value) error {
+	switch val.Kind() {
+	case reflect.Struct:
+		return validateStruct(val)
+
+	case reflect.Array, reflect.Slice:
+		return validateArrayOrSlice(val)
+
+	case reflect.Map:
+		return validateMap(val)
+
+	case reflect.Ptr:
+		return validatePointer(val)
+
+	case reflect.Bool:
+		return nil
+
+	default:
+		if val.IsZero() {
+			return fmt.Errorf("required field is empty")
+		}
+	}
+	return nil
+}
+
+func isStructPtr(fieldValue reflect.Value) bool {
+	return fieldValue.Kind() == reflect.Ptr && !fieldValue.IsNil() && fieldValue.Elem().Kind() == reflect.Struct
+}
+
+func tagContainsOmitempty(tag string) bool {
+	return strings.Contains(tag, models.OmitemptyTag)
+}
+
+func validateStruct(val reflect.Value) error {
+	for i := 0; i < val.NumField(); i++ {
+		field := val.Type().Field(i)
+		fieldValue := val.Field(i)
+
+		if (fieldValue.Kind() == reflect.Array || fieldValue.Kind() == reflect.Slice) && fieldValue.Len() > 0 {
+			err := validateValue(fieldValue)
+			if err != nil {
+				return fmt.Errorf("%s, %w", field.Name, err)
+			}
+		}
+
+		if tag, ok := field.Tag.Lookup(models.JsonTag); ok && tagContainsOmitempty(tag) {
+			if fieldValue.Kind() == reflect.Ptr || fieldValue.Kind() == reflect.Map {
+				if !fieldValue.IsZero() {
+					err := validateValue(fieldValue)
+					if err != nil {
+						return fmt.Errorf("%s, %w", field.Name, err)
+					}
+				}
+			}
+			continue
+		}
```

Comment:
What do you think about these changes? It simplifies the code by removing wrapping of if statements
```suggestion
		if (fieldValue.Kind() == reflect.Array || fieldValue.Kind() == reflect.Slice || fieldValue.Kind() == reflect.Map) && fieldValue.Len() > 0 {
			err := validateValue(fieldValue)
			if err != nil {
				return fmt.Errorf("%s, %w", field.Name, err)
			}
		}

		if tag, ok := field.Tag.Lookup(models.JsonTag); ok && tagContainsOmitempty(tag) {
			if fieldValue.Kind() == reflect.Ptr && fieldValue.IsZero()  {
			   continue
			}
			
			 err := validateValue(fieldValue)
			 if err != nil {
				return fmt.Errorf("%s, %w", field.Name, err)
			}
		}
```

Reply:
Also what do you think about adding the the func for the next code:

```go
func hasLength(val reflect.Value) bool {
  return val.Kind() == reflect.Array || val.Kind() == reflect.Slice || val.Kind() == reflect.Map
}
```

And you it inside your `validateStruct` func:
```go
...
if hasLength(fieldValue) && fieldValue.Len() > 0 {
	...
}
```

Reply:
in the first suggestion you don't consider the required pointers